saleor/discount/models.py

```diff
@@ -109,31 +110,30 @@ def get_apply_to_display(self):
             return choices[self.apply_to]
         return None
 
-    def get_fixed_discount_for(self, amount):
+    def get_discount(self):
         if self.discount_value_type == DiscountValueType.FIXED:
-            discount_price = Price(
-                net=self.discount_value, currency=settings.DEFAULT_CURRENCY)
-            discount = FixedDiscount(
-                amount=discount_price, name=smart_text(self))
-        elif self.discount_value_type == DiscountValueType.PERCENTAGE:
-            discount = percentage_discount(
-                value=self.discount_value, name=smart_text(self))
-            fixed_discount_value = amount - discount.apply(amount)
-            discount = FixedDiscount(
-                amount=fixed_discount_value, name=smart_text(self))
-        else:
-            raise NotImplementedError('Unknown discount value type')
-        if discount.amount > amount:
-            return FixedDiscount(amount, name=smart_text(self))
-        return discount
+            discount_amount = Money(
+                self.discount_value, currency=settings.DEFAULT_CURRENCY)
+            return partial(fixed_discount, discount=discount_amount)
+        if self.discount_value_type == DiscountValueType.PERCENTAGE:
+            return partial(percentage_discount, percentage=self.discount_value)
+        raise NotImplementedError('Unknown discount type')
+
+    def get_discount_amount_for(self, price):
+        discount = self.get_discount()
+        gross_amount = price.gross
+        gross_after_discount = discount(gross_amount)
+        if gross_after_discount < Money(0, currency=settings.DEFAULT_CURRENCY):
```

Comment:
`if gross_after_discount.amount < 0:`? Also wouldn't it be easier to make sure that discounts do not result in negative amounts?

Reply:
Yeah, comparing amount against zero is simpler than what I've came up with.

As for making sure that discount won't be negative, I've experimented with being safe before being sorry here, but couldn't come with much. For fixed discounts, you can check for `price.gross.amount <= self.discount_value`, and for percentage for `self.discount_value > 100`. For percentage discounts we could add check to dashboard's form to prevent nonsensical discounts like 101%. On sidenote, checking for nergative has its upsides too, custom discount types in forks would be guaranteed to never produce negatives without having to modify the check. Just adding custom type to `get_discount` would cut it.

Reply:
I meant that the actual implementations of `fixed_discount` and `percentage_discount` should never return a negative number. Otherwise all consumers of the `prices` API have to add these checks in their code.